### PR TITLE
[GraphOptimizer] Optimization to merge Transpose into MatMul

### PR DESCRIPTION
**Pull request by aturetsk**
*Description*:
GLOW uses NHWC format for convolutions and this leads to appearance of a transpose before matmuls in some classification networks if NCHW is used as input layout. This patch adds optimization to get rid of such transpose by merging it into matmul and reordering matmul's weights.
*Testing*:
Added a test in graphOptzTest.
Verified with internal classification net.
*Documentation*:
N/A

### CODE REVIEW

File: lib/Optimizer/GraphOptimizer.cpp [hunk 1]
```diff
@@ -592,6 +592,78 @@ static void mergeMatMul(Function *F) {
   }
 }
 
+/// Merge Transpose into MatMul.
```
Comment (nadavrot) [inline lib/Optimizer/GraphOptimizer.cpp]: I wonder about the reshape in the optimization. Is the reshape coming from FC node? Would it make sense to catch this pattern before lowering the FC?

File: lib/Optimizer/GraphOptimizer.cpp [hunk 2]
```diff
@@ -592,6 +592,78 @@ static void mergeMatMul(Function *F) {
...
+      // in any way.
+      auto *TN = dyn_cast<TransposeNode>(RN->getInput());
+      if (!TN || !TN->hasOneUse() || TN->getShuffle()[0] != 0) {
+        continue;
+      }
+
+      // Reshaped weight types and shuffle to be able to simply use
+      // Tensor::transpose for weight reordering.
+      ShapeVector shape, newShape;
+      llvm::SmallVector<unsigned_t, max_tensor_dimensions> shuffle;
```
Comment (nadavrot) [inline lib/Optimizer/GraphOptimizer.cpp]: Could you please add a few comments here that describe the transformation?

File: lib/Optimizer/GraphOptimizer.cpp [hunk 3]
```diff
@@ -1906,6 +1978,12 @@ void glow::optimize(Function *F, CompilationMode mode) {
 
   optimizeReshape(F);
 
+  // Merge Transpose into MatMul.
+  // Run after optimizeReshape to ensure a single Reshapes precedes MatMul.
+  // Run DCE to ensure correct number of node users.
```
Comment (nadavrot) [inline lib/Optimizer/GraphOptimizer.cpp]: Why do we need the DCE here?

Comment (aturetsk) [reply lib/Optimizer/GraphOptimizer.cpp]: Optimizations do not erase nodes, they just do 'ReplaceAllUsesWith' and relay on DCE to clean up. This corrupts number of users for nodes (after transformation we usually have a newly created optimized use and old 'dead' use), i.e. node->hasOneUse() may return false, while after DCE it'll be true.

File: lib/Optimizer/GraphOptimizer.cpp (hunk 1 above)
Comment (nadavrot) [reply lib/Optimizer/GraphOptimizer.cpp]: Great!

File: lib/Optimizer/GraphOptimizer.cpp (hunk 3 above)
Comment (nadavrot) [reply lib/Optimizer/GraphOptimizer.cpp]: Oh, wow this is a bug!  Could you please file an issue to track this bug?  Optimizations should not corrupt the graph.

File: lib/Optimizer/GraphOptimizer.cpp (hunk 1 above)
Comment (aturetsk) [reply lib/Optimizer/GraphOptimizer.cpp]: I realized I was wrong. Reshape comes even before lowering from Caffe2 importer:
https://github.com/pytorch/glow/blob/master/lib/Importer/Caffe2.cpp line 300.
So there is no much sense to move the optimization to pre-lowering stage.

**Comment by aturetsk** [comment]
Used early exit in the optimization and added more comments describing transformation.

File: lib/Optimizer/GraphOptimizer.cpp (hunk 3 above)
Comment (aturetsk) [reply lib/Optimizer/GraphOptimizer.cpp]: Created https://github.com/pytorch/glow/issues/1688

**Comment by nadavrot** [comment]
Thank you for doing this work Andrey! @aturetsk

**Comment by nadavrot** [comment]
Andrey, can we land this PR?
@aturetsk

**Comment by nadavrot** [comment]
I will land this PR to avoid merge conflict. I just realized that you don't have permissions to land pull requests. Next time please add a comment on the PR and someone will land it as soon as the PR is approved. Sorry about the delay.

**Pull request merged**